### PR TITLE
Shubhadapaithankar/aro-9323-cirp-mvp

### PR DESCRIPTION
**Pull request by shubhadapaithankar**
### Which issue this PR addresses:
Fixes ARO-9323
This PR introduces a minimum viable PR validation pipeline that leverages Docker instead of Podman. It ensures consistency across both local development and the Azure DevOps (ADO) pipeline by running `make ci ` automatically.

### What this PR does / why we need it:

- Streamlined ADO Pipeline: Sets up an Azure DevOps pipeline to streamline the PR validation process.
- Automated Build and Test: Runs make ci in the pipeline to automate building, validating, and testing the RP container image using Docker.
- Image Push to ACR: Automates the process of pushing validated Docker images to the Azure Container Registry (ACR) for development, ensuring a consistent and efficient workflow.
- Unified Environment: Replaces Podman with Docker to ensure local development and CI pipelines are aligned, avoiding discrepancies between the two.
- Coverage Extraction: Extracts coverage reports from containerized environments using Docker, ensuring accurate test coverage data.

###  Key Features
- Builds faster than the previous method (~16 minutes vs ~19 minutes; an 18% improvement is significant!).
- Performs additional checks earlier in the process (e.g., FIPS validation, NPM audit).
- Preserves the artifact by pushing the image, allowing you to pull and debug locally if your PR fails later.
- Streamlines everything into a single, consistent process.

### Test plan for issue:
- This PR includes automated unit tests as part of the `make ci `process.This ensures that code changes are validated against predefined tests.
- The pipeline also runs integration/e2e tests to validate the changes in a realistic environment.
### Is there any documentation that needs to be updated for this PR?
N/A.
### How do you know this will function as expected in production?

### CODE REVIEW

File: .pipelines/ci.yml
Comment (SudoBrendan) [inline .pipelines/ci.yml:183]: idea: instead of using `awk` on a log to get an image tag (and thus also hiding the log from view in the build), then retagging that image, we should be able to save both steps, and just override `LOCAL_AZ_EXT_ARO_IMAGE` envvar to set it to the image name we need, right? If our issue is that we need to know what `VERSION` is, let's just make a Makefile target like `make version` to grab that data directly rather than parsing a log.

File: .pipelines/ci.yml
Comment (SudoBrendan) [inline .pipelines/ci.yml:282]: can we move this logic into the Makefile `ci-rp` target, so that we can also view this file on local builds?

File: .pipelines/ci.yml
Comment (SudoBrendan) [inline .pipelines/ci.yml:243]: idea: these commands appear to be common. Can we extract them into a template so we can use them consistently, everywhere?

Comment (shubhadapaithankar) [reply .pipelines/ci.yml]: I have created a dedicated template, template-podman.yml, which consolidates these commands, including the installation of Podman, starting the Podman service, and validation steps. This template is now used consistently across the pipeline.

File: Dockerfile.ci-rp
Comment (SudoBrendan) [inline Dockerfile.ci-rp:74]: We won't want to do this. Instead, let's [extract](https://unix.stackexchange.com/questions/331645/extract-file-from-docker-image) the `cover.out` from our tagged rp-build image in the Makefile at the end of `ci-rp`

File: .pipelines/templates/template-podman.yml
Comment (SudoBrendan) [inline .pipelines/templates/template-podman.yml:84]: I notice we're using `--remote --url=tcp://localhost:8888` throughout this - can we make that a standard parameter like we did in the Makefile? Alternatively, we may be able to configure podman's default configuration to point to that connection?

File: .pipelines/templates/template-podman.yml
Comment (SudoBrendan) [inline .pipelines/templates/template-podman.yml:173]: this is the only usage of `az` in this script - is this step necessary (my guess is "no" since we are building in the `ubi8` container, which is destroyed on each build iteration, or is that false?)? If it isn't we could remove `az` from this build process.

File: .pipelines/templates/template-podman.yml
Comment (SudoBrendan) [inline .pipelines/templates/template-podman.yml:165]: This looks like a near-dupe of what `make-ci` does. Instead of printing the contents to stdout, let's use the `task: PublishCodeCoverageResults@1` as used in the noncontainerized build process. This integrates our unit test results with ADO directly so we get a UI representation of passes/failures and can detect things like flaky tests over time/etc.

**Review by SudoBrendan — CHANGES_REQUESTED** [review]
Some initial comments - the big one is using the test publishing step rather than `cat`

File: .pipelines/templates/template-podman.yml
Comment (SudoBrendan) [inline .pipelines/templates/template-podman.yml]: q: interesting - do we need `libicu`? That package doesn't sound familiar to me - just curious what it is?

File: .pipelines/ci.yml
Comment (SudoBrendan) [inline .pipelines/ci.yml:103]: while we're here - let's fix a quick bug - this step needs to run in `target: golang`

File: .pipelines/templates/template-podman.yml
Comment (SudoBrendan) [inline .pipelines/templates/template-podman.yml:147]: I think we can improve this. The fact we have to install `go` in this stuff here in the pipeline is telling me that `ci-rp` is not doing the same thing as the old build, where we had `make unit-test-go` -> `make xmlcov` -> publish results and coverage files:

```
        - script: |
            set -xe
            make unit-test-go
          displayName: 🧪 Run Golang unit tests
          target: golang

        - task: PublishTestResults@2
          displayName: 📊 Publish tests results
          inputs:
            testResultsFiles: $(System.DefaultWorkingDirectory)/**/report.xml
          condition: succeededOrFailed()
          // **BUG: this needs to run in `target:golang` - how is this even working? (or is it?)

        - script: |
            make xmlcov
          displayName: ⚙️ Process Reports
          condition: succeededOrFailed()
          target: golang

        - task: PublishCodeCoverageResults@1
          displayName: 📈 Publish code coverage
          inputs:
            codeCoverageTool: Cobertura
            summaryFileLocation: $(System.DefaultWorkingDirectory)/**/coverage.xml
            failIfCoverageEmpty: false
          condition: succeededOrFailed()
          target: golang
```

What I'd like to see in the new pipeline is effectively a one line code change (in actuality, we know that it's not "just" one line because of podman setup/etc ... but `make ci-rp` should give us exactly the same local files we expected before. We should be able to streamline the process into: `make ci-rp` -> publish coverage files.

```
        - script: |
            set -xe
            make ci-rp
          displayName: 🧪 Run Golang unit tests
          target: golang

          // rest is the same
```

Can we update the Dockerfile and "extract" steps in the Makefile to get us all the same local files we need, rather than rolling a new process here in the pipe?

**Review by SudoBrendan — CHANGES_REQUESTED** [review]
Good progress - let's simplify the test coverage publishing a bit before merge.

File: .pipelines/ci.yml
Comment (SudoBrendan) [inline .pipelines/ci.yml:137]: If I'm reading this correctly `repository` is the same for all images. Let's fix that so that each image build is a unique name that is representative of the artifact (should probably be in sync with the Makefile `LOCAL_*_IMAGE` names).

File: .pipelines/ci.yml
Comment (shubhadapaithankar) [reply .pipelines/ci.yml:103]: Added target: golang to the PublishTestResults step.

File: .pipelines/templates/template-podman.yml
Comment (shubhadapaithankar) [reply .pipelines/templates/template-podman.yml]: removed the libicu

File: .pipelines/ci.yml
Comment (shubhadapaithankar) [reply .pipelines/ci.yml]: Updated the pipeline to use unique repository names by referencing LOCAL_*_IMAGE variables from the Makefile. This ensures image names are consistent and in sync with the Makefile definitions.

File: .pipelines/templates/template-podman.yml
Comment (SudoBrendan) [inline .pipelines/templates/template-podman.yml:116]: idea: replace this by just setting `LOCAL_*_IMAGE` and then make will do that for you

Comment (shubhadapaithankar) [reply .pipelines/templates/template-podman.yml]: Done

File: .pipelines/templates/template-podman.yml
Comment (shubhadapaithankar) [reply .pipelines/templates/template-podman.yml]: Please see the latest code with fix

File: .pipelines/templates/template-docker.yml
Comment (SudoBrendan) [reply .pipelines/templates/template-docker.yml:66]: Alright - so we discovered due to cross-tenant auth issues, we can't use the `command: login` - so that's out, but we should be able to use current auth into ACR + `Docker@2` for a `buildAndPush` call to remove a bit of `sh` here, I think?

File: .pipelines/templates/template-docker.yml
Comment (SudoBrendan) [inline .pipelines/templates/template-docker.yml:70]: we can skip this by modifying what `repository` is above to be `${ACR_FQDN}/${{ parameters.repository }}`

File: Makefile
Comment (SudoBrendan) [inline Makefile:417]: I think removing `--noheading` and doing `head -n 1` will only get the header for the images. This will need to be modified.

Comment (SudoBrendan) [reply Makefile:416]: nevermind - it looks like `--format` doesn't print headings (?) https://docs.docker.com/reference/cli/docker/image/ls/#format

File: .pipelines/templates/template-acr-push.yml
Comment (SudoBrendan) [inline .pipelines/templates/template-acr-push.yml:19]: We should document in a comment why this template is even necessary (the answer is we want the ability to push images across Azure tenants, and MSI won't cut it) so that future maintainers understand why we didn't go the "easy" route of configuring `Docker@2`'s `push` command.

File: hack/e2e/utils.sh
Comment (SudoBrendan) [inline hack/e2e/utils.sh:1]: Are these changes used in this PR anymore or can they be removed?

**Review by SudoBrendan — APPROVED** [review]
Ultimately LGTM - maybe a few more places for us to add documentation, but otherwise this looks really clean and clear as to what is happening, and most importantly: it is _consistent_. Fantastic work :)

File: hack/e2e/utils.sh
Comment (SudoBrendan) [inline hack/e2e/utils.sh:1]: whoops, I was unclear - we want this file as-is in master instead of the changes that were previously here :) this is still in-use for current E2E

Comment (shubhadapaithankar) [reply hack/e2e/utils.sh:1]: We are not using this anymore, file has been removed.

**Review by SudoBrendan — CHANGES_REQUESTED** [review]
Just a "whoops" fix needed :)

**Review by SudoBrendan — APPROVED** [review]

File: hack/e2e/utils.sh
Comment (shubhadapaithankar) [reply hack/e2e/utils.sh:1]: I've resolved it and reverted the file to match what's in master.

File: .pipelines/templates/template-acr-push.yml
Comment (shubhadapaithankar) [reply .pipelines/templates/template-acr-push.yml:22]: I've added the comment explaining why this template is necessary for cross-tenant authentication.

**Review by slawande2 — APPROVED** [review]
LGTM!

File: .pipelines/templates/template-acr-push.yml
Comment (SudoBrendan) [inline .pipelines/templates/template-acr-push.yml:60]: nit: we could maybe simplify this template by ensuring that at `docker build` time, we create an image that we expect including the ACR domain/image name/tag (e.g. by setting the `LOCAL_*_IMAGE` variables Makefile references. That would mean this template accepts that fully-qualified image then authenticates and pushes without additional checks or `docker` commands.

File: Makefile
Comment (SudoBrendan) [inline Makefile:404]: q: how is this used and why is it important?

**Review by SudoBrendan — APPROVED** [review]
LGTM, a few comments/questions but they wouldn't stop merge.

File: Makefile
Comment (SudoBrendan) [inline Makefile:421]: I found Waldo! This second `--no-cache` is why the builds take so long! We should use the cache on the second build, or use `--cache-from=$(LOCAL_ARO_RP_BUILD_IMAGE):$(VERSION)` to halve the build times.

File: Makefile
Comment (shubhadapaithankar) [reply Makefile:404]: This isn't necessary anymore; I had added it for testing purposes, but we can remove it now. Apologies for any confusion

File: Makefile
Comment (shubhadapaithankar) [reply Makefile]: Thanks for catching that! I've updated the Makefile to use --cache-from=$(LOCAL_ARO_RP_BUILD_IMAGE):$(VERSION) for the second build instead of --no-cache, which should speed up the build process by reusing cached layers

**Review by slawande2 — DISMISSED** [review]
LGTM.

File: Makefile
Comment (SudoBrendan) [inline Makefile:416]: I think that we will want `$(LOCAL_ARO_RP_BUILD_IMAGE)` instead of `$(REGISTRY)/$(LOCAL_ARO_RP_BUILD_IMAGE)` - we want to use the local images as a cache - the one we just built, rather than something remote.

Comment (shubhadapaithankar) [reply Makefile]: Yeah right ,Thanks for the feedback! I have updated the code to use `$(LOCAL_ARO_RP_BUILD_IMAGE) `instead of `$(REGISTRY)/$(LOCAL_ARO_RP_BUILD_IMAGE)`

**Review by SudoBrendan — CHANGES_REQUESTED** [review]
A few pieces of dead code can be removed in the template :)

File: .pipelines/templates/template-acr-push.yml
Comment (SudoBrendan) [inline .pipelines/templates/template-acr-push.yml:12]: in all of the uses of these parameters, they are the same. Can we consolidate?

File: .pipelines/templates/template-acr-push.yml
Comment (SudoBrendan) [inline .pipelines/templates/template-acr-push.yml:14]: this appears to be dead code - let's remove it and all references.

File: .pipelines/templates/template-acr-push.yml
Comment (shubhadapaithankar) [reply .pipelines/templates/template-acr-push.yml]: Thanks for pointing me .- Removed unused parameters: `dockerFilePath`, `jobName`, and `imageTag` from template-acr-push.yml.

File: .pipelines/templates/template-acr-push.yml
Comment (shubhadapaithankar) [reply .pipelines/templates/template-acr-push.yml]: - Removed unused parameters: `dockerFilePath`, `jobName`, and `imageTag` from template-acr-push.yml.

File: .pipelines/templates/template-acr-push.yml
Comment (shubhadapaithankar) [reply .pipelines/templates/template-acr-push.yml]: - Consolidated `repository` and `imageRepoName` parameters into a single `repository` parameter.

**Review by SudoBrendan — APPROVED** [review]
Beautiful code. It builds faster than the old way, can be executed locally, and performs additional checks that were normally put off until later (e.g. FIPS validation) to simplify everything into a single, consistent process. LGTM!

**Review by slawande2 — APPROVED** [review]

**Review by kevinobriendotca — APPROVED** [review]
LGTM

**Pull request merged**
